src/porepy/models/initial_condition.py

```diff
@@ -0,0 +1,117 @@
+
+            class ICPrimary(InitialConditionMixin):
+
+                def set_initial_values_primary_variables(self) -> None:
+                    super().set_initial_values_primary_variables()
+
+                    for sd in self.mdg.subdomains():
+                        self.equation_system.set_variable_values(
+                            self.initial_x(sd),
+                            [self.x([sd])],
+                            iterate_index=0,
+                        )
+
+                def initial_x(self, sd: pp.Grid) -> np.ndarray:
+                    # proceed to return some value
+                    ...
+
+            class ICSecondary(InitialConditionMixin):
+
+                def initial_condition(self) -> None:
+                    super().initial_condition()
+
+                    for sd in self.mdg.subdomains():
+                        self.equation_system.set_variable_values(
+                            self.initial_y(sd),
+                            [self.y([sd])],
+                            iterate_index=0,
+                        )
+
+                def initial_y(self, sd: pp.Grid) -> np.ndarray:
+                    x = self.x([sd]).value(self.equation_system)
+                    # proceed to return some value depending on x
+                    ...
+
+            class MyIC(ICSecondary, ICPrimary):
+                ...
+
+        Notice that in all initialization methods, ``super()`` is called first. For the
+        model using ``MyIC``, because of the order of inheritance, the code in
+        ``ICSecondary`` is executed first, which in return executes the code of
+        ``ICPrimary`` first. I.e., the update order is the reverse order in the
+        inheritance tree. Thus, when using this approach, the IC update for ``y`` can
+        reliably fetch the latest values for ``x`` on the boundary.
+
+        Notice also, that ``set_initial_values_primary_variables`` has also a
+        ``super()`` call on top. This makes it compatible in the case of a third,
+        primary variable which should be initialized in the same sub-routine as ``x``.
+
+    """
+
+    def initial_condition(self) -> None:
+        """Attaches to the implementation of ``initial_condition`` of the solution
+        strategy.
+
+        Calls the methods :meth:`set_initial_values_primary_variables` and copies values
+        stored at iterate index 0 to all other time and iterate indices.
+
+        """
+        # First call super and let the solution strategy handle the initialization of
+        # everything (likely with zeros).
+        # Then run the functionality given by this class.
+
+        # Mypy complains about the parent (the protocol) having a trivial body.
+        # We ignore the safe-super check here, but do not compromise safety by
+        # explicitly checking the inheritance tree.
+        if isinstance(self, pp.SolutionStrategy):
+            super().initial_condition()  # type:ignore[safe-super]
```

Comment:
@IvarStefansson, can we please have your meaning on the following: Should the SolutionStrategy.initial_condition() be kept, as is done here, or if we can replace it with this new concept? This may require that you have a short look at the various InitialCondition mixins (search class InitialCondition, I believe) that have been introduced in the models, but not more - so if you feel you need to read deeper, I have most likely underestimated something.

Reply:
I am not sure, there is both `InitialConditionsSinglePhaseFlow` and `InitialConditionsEnergy` in the respective modules.

Reply:
Update: @vlipovac pointed out that except from the tests of the compositional model introduced here, all tests use the old version. My suggestion is that we keep the old initialization for now, but try to convert the existing tests soon after this PR is merged. This could be a nice user project.

@IvarStefansson: Let me know what you think.

Reply:
> Sounds like a good solution. Could you make a issue with the user group tag, please, @keileg?

#1324.

Reply:
@vlipovac `ic_values` is approved.